### DISCUSSION TITLE
useState 利用 lazy initilizer 沒辦法異步函數的返回值設為 state 的初始值

### DISCUSSION THREAD

**Discussion by Wangpoching** [Lidemy/forum#236] [前端相關問題] [answered]
為了要防止 render 完又因為 useEffect 的 setValue 重新 render 頁面，所以把 fetch 到的資料直接在 useSatate 的時候設定初始值。 **但是初始值即使已經使用 await 仍然是 promise (已有結果的 promise)**

[codesandbox](https://codesandbox.io/s/peaceful-varahamihira-d1sut)

如果 codeSandBox 不能看程式碼如下:
```js
export default function App() {
  const [messages, setMessage] = useState(async () => {
    return  (
      await fetch(API_ENDPOINT)
        .then((data) => data.json())
        .then((data) => data)
    )
  })
  console.log('messages',messages)
  return (
     <div>hello</div>
  )
}
```

**Comment by aszx87410** [comment, marked as answer]
lazy initializer 應該本來就不支援非同步
因為你想想，如果支援的話，那 state 要等非同步結果回來才能確定值，那等待的過程該怎麼辦？畫面只能一片空白，這樣使用者體驗只會更差

所以你不管怎樣都無法避免 rerender，而且 rerender 沒有你想像中的可怕，大多數狀況不刻意去避開也可以
